### PR TITLE
cluster: handle region after report split

### LINKED ISSUE #4157: Update Region Meta in advance during ReportBatchSplit
### Describe your feature request related problem

After the split occurs in the Region, Heartbeat is relied on to update the Region Meta. The Range hole will be generated during the update, and the Region corresponding to the Range cannot be queried in a short time.

When the PD heartbeat processing is not fast enough, the problem will be magnified.

### Describe the feature you'd like

Update Region Meta in advance during `ReportBatchSplit`

### PR DESCRIPTION
**Pull request by bufferflies**
Signed-off-by: bufferflies <1045931706@qq.com>

### What problem does this PR solve?
Issue Number: close #4157 and close https://github.com/tikv/tikv/issues/15210

### What is changed and how it works?
put region after batch report split 
### Check List

Tests

- Unit test
- Manual test

```
// test sql
use test;drop table if exists t2;create table t2 (a int, b varchar(10), unique index idx_a(a)) partition by hash (a) partitions 3;insert into t2 values (1,'a'), (11, 'h'), (22, 'o');show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2
regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;show table t2 regions;


no any err `PD returned no region`
```

Code changes

Side effects

Related changes

### Release note

```release-note
handle region after report split
```

### CODE REVIEW

**Review by lhy1024 — COMMENTED** [review]
ci failed

File: server/grpc_service.go [hunk 1]
```diff
@@ -1266,10 +1266,25 @@ func (s *GrpcServer) GetRegion(ctx context.Context, request *pdpb.GetRegionReque
...
+	var region *core.RegionInfo
+	// allow region miss temporarily if this key can't be found in the region tree.
+	for retry := 0; retry <= 10; retry++ {
+		region = rc.GetRegionByKey(request.GetRegionKey())
+		if region != nil {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			break
```
Comment (CabinfeverB) [inline server/grpc_service.go:1278]: You should name the `for` loop, and break this loop. Now, it just break out the `select`.

File: server/grpc_service.go [hunk 2]
```diff
@@ -1300,7 +1315,22 @@ func (s *GrpcServer) GetPrevRegion(ctx context.Context, request *pdpb.GetRegionR
...
+	var region *core.RegionInfo
+	// allow region miss temporarily if this key can't be found in the region tree.
+	for retry := 0; retry <= 10; retry++ {
+		region = rc.GetPrevRegionByKey(request.GetRegionKey())
+		if region != nil {
+			break
+		}
+		select {
+		case <-ctx.Done():
+			break
```
Comment (CabinfeverB) [inline server/grpc_service.go:1327]: ditto

File: server/cluster/cluster_worker.go [hunk 3]
```diff
@@ -166,19 +165,7 @@ func (c *RaftCluster) HandleAskBatchSplit(request *pdpb.AskBatchSplitRequest) (*
 }
 
 func (c *RaftCluster) checkSplitRegion(left *metapb.Region, right *metapb.Region) error {
```
Comment (CabinfeverB) [inline server/cluster/cluster_worker.go:167]: It is completely replaced by `processRegionSplit`, can remove it.

File: server/cluster/cluster.go [hunk 4]
```diff
@@ -1082,6 +1082,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
 	return nil
 }
 
+func (c *RaftCluster) processRegionSplit(regions []*metapb.Region) []error {
```
Comment (CabinfeverB) [inline server/cluster/cluster.go:1085]: How about put it on `cluster_worker.go`

File: server/cluster/cluster.go [hunk 5]
```diff
@@ -1082,6 +1082,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
...
+				}
+			}
+		}
+		region := core.NewRegionInfo(region, leader)
+		overlaps, perr := c.core.AtomicCheckAndPutRegion(region)
+		if perr != nil {
+			errList = append(errList, perr)
+			continue
+		}
+		for _, item := range overlaps {
```
Comment (CabinfeverB) [inline server/cluster/cluster.go:1118]: This loop also appears in `processRegionHeartbeat`, how about extracting it?

File: server/cluster/cluster.go [hunk 6]
```diff
@@ -1082,6 +1082,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
...
+			if c.regionStats != nil {
+				c.regionStats.ClearDefunctRegion(item.GetID())
+			}
+			c.labelLevelStats.ClearDefunctRegion(item.GetID())
+			c.ruleManager.InvalidCache(item.GetID())
+		}
+		regionUpdateCacheEventCounter.Inc()
+		if c.storage != nil {
```
Comment (CabinfeverB) [inline server/cluster/cluster.go:1126]: ditto

**Review by CabinfeverB — COMMENTED** [review]
Can we add some comment to show the difference processRegionSplit needs to do less than processRegionHeartbeat

File: server/cluster/cluster.go (hunk 5 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: server/cluster/cluster.go (hunk 6 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: server/cluster/cluster.go [hunk 7]
```diff
@@ -1082,6 +1082,72 @@ func (c *RaftCluster) processRegionHeartbeat(region *core.RegionInfo) error {
...
+	}
+	total := len(regions) - 1
+	regions[0], regions[total] = regions[total], regions[0]
+	leaderStoreID := uint64(0)
+	if r := c.core.GetRegion(regions[0].GetId()); r != nil {
+		leaderStoreID = r.GetLeader().GetStoreId()
+	}
+	errList := make([]error, 0, total)
+	for _, region := range regions {
+		// It should update rightmost because it can override by left.
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/cluster/cluster.go]: no, only ensure the origin region first update.

File: server/cluster/cluster.go (hunk 4 above)
Comment (bufferflies) [reply server/cluster/cluster.go]: fixed

File: server/cluster/cluster_worker.go [hunk 8]
```diff
@@ -41,6 +41,49 @@ func (c *RaftCluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
...
+		return []error{errors.New("origin region no leader")}
+	}
+	errList := make([]error, 0, total)
+	for _, region := range regions {
+		if len(region.GetPeers()) == 0 {
+			errList = append(errList, errors.New(fmt.Sprintf("region:%d has no peer", region.GetId())))
+			continue
+		}
+		// region split initiator store will be leader with a high probability
+		leader := region.Peers[0]
```
Comment (rleungx) [inline server/cluster/cluster_worker.go:67]: Will it cause the balance leader to wrong place?

Comment (bufferflies) [reply server/cluster/cluster_worker.go:67]: good catch.

File: server/cluster/cluster_worker.go [hunk 9]
```diff
@@ -41,6 +41,49 @@ func (c *RaftCluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
...
+// so it doesn't consider the leader and hot statistics.
+func (c *RaftCluster) ProcessRegionSplit(regions []*metapb.Region) []error {
+	if err := c.checkSplitRegions(regions); err != nil {
+		return []error{err}
+	}
+	total := len(regions) - 1
+	regions[0], regions[total] = regions[total], regions[0]
+	leaderStoreID := uint64(0)
+	if r := c.core.GetRegion(regions[0].GetId()); r != nil {
+		leaderStoreID = r.GetLeader().GetStoreId()
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/cluster/cluster_worker.go:55]: it will return 0  if leader not found

File: server/cluster/cluster_worker.go (hunk 8 above)
Comment (bufferflies) [reply server/cluster/cluster_worker.go:67]: I will mark this store as `split store` and balance region scheduler doesn't pick this store as target store.

**Comment by HuSharp** [comment]
pls remove `rebase` label, thx!

File: pkg/schedule/plan/status.go [hunk 10]
```diff
@@ -72,6 +72,7 @@ const (
 	StatusStoreLowSpace = iota + 500
 	// StatusStoreNotExisted represents the store cannot be found in PD.
 	StatusStoreNotExisted
+	StatusStoreSplitRegion
```
Comment (HuSharp) [inline pkg/schedule/plan/status.go:75]: maybe need to add name as same as `StatusStoreNotExisted`

File: pkg/syncer/client.go [hunk 11]
```diff
@@ -194,15 +194,15 @@ func (s *RegionSyncer) StartSyncWithLeader(addr string) {
...
+					changed := regionGuide(region, origin)
 					overlaps := bc.PutRegion(region)
 
 					if hasBuckets {
 						if old := origin.GetBuckets(); buckets[i].GetVersion() > old.GetVersion() {
 							region.UpdateBuckets(buckets[i], old)
 						}
 					}
-					if saveKV {
+					if changed.SaveCache {
```
Comment (HuSharp) [inline pkg/syncer/client.go:205]: why changed it

File: server/grpc_service.go [hunk 12]
```diff
@@ -1266,10 +1266,24 @@ func (s *GrpcServer) GetRegion(ctx context.Context, request *pdpb.GetRegionReque
...
+		region = rc.GetRegionByKey(request.GetRegionKey())
+		if region != nil {
+			break retryLoop
+		}
+		select {
+		case <-ctx.Done():
+			break retryLoop
+		case <-time.After(10 * time.Millisecond):
+		}
+	}
```
Comment (HuSharp) [inline server/grpc_service.go:1282]: How about abstracting it?

Comment (bufferflies) [reply server/grpc_service.go:1282]: not fix in this pr , will add in other pr.

File: pkg/schedule/plan/status.go (hunk 10 above)
Comment (bufferflies) [reply pkg/schedule/plan/status.go:75]: good catch , has fixed.

File: pkg/syncer/client.go (hunk 11 above)
Comment (bufferflies) [reply pkg/syncer/client.go]: good catch, fixed it.

File: pkg/schedule/schedulers/balance_leader.go [hunk 13]
```diff
@@ -489,6 +490,15 @@ func (l *balanceLeaderScheduler) transferLeaderIn(solver *solver, collector *pla
 		balanceLeaderNoLeaderRegionCounter.Inc()
 		return nil
 	}
+
+	if solver.Source.IsSplitStore() {
+		log.Debug("source store split region in one minutes",
```
Comment (HuSharp) [inline pkg/schedule/schedulers/balance_leader.go:495]: Is it better to print `splitStoreWait` other than `one minutes`?

**Review by HuSharp — APPROVED** [review]
rest LGTM!

File: pkg/schedule/filter/filters.go [hunk 14]
```diff
@@ -332,6 +332,8 @@ type StoreStateFilter struct {
 	// If it checks failed, the operator will be put back to the waiting queue util the limit is available.
 	// But the scheduler should keep the same with the operator level.
 	OperatorLevel constant.PriorityLevel
+	// set true if not allow balance leader from this store
+	ForbidSplit bool
```
Comment (rleungx) [inline pkg/schedule/filter/filters.go:336]: How about changing to another name? It sounds strange.

File: server/cluster/cluster_worker.go [hunk 15]
```diff
@@ -41,6 +45,58 @@ func (c *RaftCluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
...
+		region := core.NewRegionInfo(region, leader)
+		changed := &core.RegionChanged{
+			IsNew: true, SaveKV: true, SaveCache: true, NeedSync: true,
+		}
+		if err := c.SaveRegion(region, changed); err != nil {
+			errList = append(errList, err)
+		}
+	}
+	// If the number of regions exceeds the threshold, update the last split time.
+	if len(regions) >= maxSplitThreshold {
```
[the remark this answers is not in the document]
Comment (bufferflies) [reply server/cluster/cluster_worker.go:93]: only affect this store in balance leader scheduler.

File: pkg/core/store.go [hunk 16]
```diff
@@ -539,6 +541,11 @@ func (s *StoreInfo) NeedAwakenStore() bool {
 	return s.GetLastHeartbeatTS().Sub(s.lastAwakenTime) > awakenStoreInterval
 }
 
+// IsSplitStore checks if there are some region are splitted in this store.
+func (s *StoreInfo) IsSplitStore() bool {
```
Comment (nolouch) [inline pkg/core/store.go:545]:
```suggestion
func (s *StoreInfo) HasRecentlySplitRegions() bool {
```

File: pkg/schedule/filter/filters.go (hunk 14 above)
Comment (bufferflies) [reply pkg/schedule/filter/filters.go]: has fixed as `ForbidSplit RecentlySplitRegions `

File: pkg/core/store.go (hunk 16 above)
Comment (bufferflies) [reply pkg/core/store.go]: fixed

File: pkg/core/region.go [hunk 17]
```diff
@@ -719,57 +725,54 @@ func GenerateRegionGuideFunc(enableLog bool) RegionGuideFunc {
...
 				return
 			}
 			if region.GetReplicationStatus().GetState() != replication_modepb.RegionReplicationState_UNKNOWN &&
 				(region.GetReplicationStatus().GetState() != origin.GetReplicationStatus().GetState() ||
 					region.GetReplicationStatus().GetStateId() != origin.GetReplicationStatus().GetStateId()) {
-				saveCache = true
+				changed.SaveCache = true
 				return
 			}
-			// Do not save to kv, because 1) flashback will be eventually set to
```
Comment (CabinfeverB) [inline pkg/core/region.go:768]: Why remove comments

Comment (bufferflies) [reply pkg/core/region.go:768]: fixed

File: pkg/schedule/schedulers/balance_leader.go [hunk 18]
```diff
@@ -57,13 +57,14 @@ const (
...
-	balanceLeaderSkipCounter             = schedulerCounter.WithLabelValues(BalanceLeaderName, "skip")
-	balanceLeaderNewOpCounter            = schedulerCounter.WithLabelValues(BalanceLeaderName, "new-operator")
+	balanceLeaderScheduleCounter           = schedulerCounter.WithLabelValues(BalanceLeaderName, "schedule")
+	balanceLeaderNoLeaderRegionCounter     = schedulerCounter.WithLabelValues(BalanceLeaderName, "no-leader-region")
+	balanceLeaderRegionHotCounter          = schedulerCounter.WithLabelValues(BalanceLeaderName, "region-hot")
+	balanceLeaderNoTargetStoreCounter      = schedulerCounter.WithLabelValues(BalanceLeaderName, "no-target-store")
+	balanceLeaderNoFollowerRegionCounter   = schedulerCounter.WithLabelValues(BalanceLeaderName, "no-follower-region")
+	balanceLeaderSkipCounter               = schedulerCounter.WithLabelValues(BalanceLeaderName, "skip")
+	balanceLeaderNewOpCounter              = schedulerCounter.WithLabelValues(BalanceLeaderName, "new-operator")
+	balanceLeaderStoreRecentlySplitRegions = schedulerCounter.WithLabelValues(BalanceLeaderName, "source-recently-split-regions")
```
Comment (CabinfeverB) [inline pkg/schedule/schedulers/balance_leader.go:67]: I think the metrics are a bit confusing here, when the source store is filtered, it is counted in 'filterCounter', but in 'transferLeaderIn', it is counted in 'schedulerCounter'.  And the label name are also different.

**Review by CabinfeverB — COMMENTED** [review]
Mostly LGTM

File: pkg/schedule/schedulers/balance_leader.go (hunk 18 above)
Comment (bufferflies) [reply pkg/schedule/schedulers/balance_leader.go]: fixed by add region filter to filter out the no splitted store.

File: pkg/schedule/schedulers/balance_leader.go [hunk 19]
```diff
@@ -486,7 +486,8 @@ func (l *balanceLeaderScheduler) transferLeaderOut(solver *solver, collector *pl
 // the worst follower peer and transfers the leader.
 func (l *balanceLeaderScheduler) transferLeaderIn(solver *solver, collector *plan.Collector) *operator.Operator {
 	solver.Region = filter.SelectOneRegion(solver.RandFollowerRegions(solver.TargetStoreID(), l.conf.Ranges),
-		nil, filter.NewRegionPendingFilter(), filter.NewRegionDownFilter())
+		nil, filter.NewRegionPendingFilter(), filter.NewRegionDownFilter(),
+		filter.NewStoreRecentlySplitFilter(solver.GetStores()))
```
Comment (CabinfeverB) [inline pkg/schedule/schedulers/balance_leader.go:490]: we can avoid to create a new filter every time

**Review by CabinfeverB — COMMENTED** [review]
rest LGTM

File: pkg/schedule/schedulers/balance_leader.go [hunk 20]
```diff
@@ -184,6 +183,7 @@ func newBalanceLeaderScheduler(opController *operator.Controller, conf *balanceL
 		&filter.StoreStateFilter{ActionScope: s.GetName(), TransferLeader: true, ForbidRecentlySplitRegions: true, OperatorLevel: constant.High},
 		filter.NewSpecialUseFilter(s.GetName()),
 	}
+	s.regionFilters = filter.NewStoreRecentlySplitFilter(opController.GetCluster().GetStores())
```
Comment (CabinfeverB) [inline pkg/schedule/schedulers/balance_leader.go:186]: It should be created in `Schedule()`

Comment (bufferflies) [reply pkg/schedule/schedulers/balance_leader.go]: good catch ， has fixed

File: server/cluster/cluster_worker.go [hunk 21]
```diff
@@ -41,6 +45,58 @@ func (c *RaftCluster) HandleRegionHeartbeat(region *core.RegionInfo) error {
...
+					leader = peer
+					break
+				}
+			}
+		}
+		region := core.NewRegionInfo(region, leader)
+		changed := &core.RegionChanged{
+			IsNew: true, SaveKV: true, SaveCache: true, NeedSync: true,
+		}
+		if err := c.SaveRegion(region, changed); err != nil {
```
Comment (nolouch) [inline server/cluster/cluster_worker.go:88]: What if the new split region has been heartbeat to pd then the  receive the report split? I think we need check the `chagned` should it be saved.

Comment (bufferflies) [reply server/cluster/cluster_worker.go:88]: It can't be hapend. Tikv send this request before the region init.

**Review by CabinfeverB — APPROVED** [review]

**Comment by HuSharp** [comment]
please fixed `GenerateRegionGuideFunc` in `server/cluster.go`

File: server/cluster/cluster_worker.go (hunk 21 above)
Comment (nolouch) [reply server/cluster/cluster_worker.go:88]: @bufferflies  network is unexpected. and they are not in the same stream.

**Pull request merged**